text/0031-corejam.md

```diff
@@ -0,0 +1,644 @@
+
+There are two kinds of Authorization corresponding to the two kinds of Coretime which are sold by Polkadot (see RFC#0001). An Authorization for usage of Instantaneous Coretime consists of a self-contained Signature of an account which own enough Instantaneous Coretime Credit in order to purchase a block of Coretime at the current rate signing a payload of the Work Package hash.
+
+RcVGs run the risk of a credit owner not having the credit at the point of eventual payment (in the Join stage, later), in which case the RcVG will not be rewarded. Credit may never be withdrawn, therefore RcVGs can safely accept a block if and only if the Credit account contains a balance of at least the product of the number of Cores assigned to IC, the price per IC core per block and the number of blocks behind the head of the finalized chain which the RcVG currently may be.
+
+An Authorization for usage of Bulk Coretime is more sophisticated. We introduce the concept of an *Authorizer* procedure, which is a piece of logic stored on-chain to which Bulk Coretime may be assigned. Assigning some Bulk Coretime to an Authorizer implies allowing any Work Package which passes that authorization process to utilize that Bulk Coretime in order to be submitted on-chain. It controls the circumstances under which RcVGs may be rewarded for evaluation and submission of Work Packages (and thus what Work Packages become valid to submit onto Polkadot). Authorization logic is entirely arbitrary and need not be restricted to authorizing a single collator, Work Package builder, parachain or even a single Work Class.
+
+An *Authorizer* is a parameterized procedure:
+
+```rust
+type CodeHash = [u8; 32];
+type AuthParamSize = ConstU32<1024>;
+type AuthParam = BoundedVec<u8, AuthParamSize>;
+struct Authorizer {
+    code_hash: CodeHash,
+    param: AuthParam,
+}
+```
+
+The `code_hash` of the Authorizer is assumed to be the hash of some code accessible in the Relay-chain's Storage pallet. The procedure itself is called the *Authorization Procedure* (`AuthProcedure`) and is expressed in this code (which must be capable of in-core VM execution). Its entry-point prototype is:
+
+```rust
+fn is_authorized(param: &AuthParam, package: &WorkPackage, core_index: CoreIndex) -> bool;
+```
+
+If the `is_authorized` function overruns the system-wide limit or panicks on some input, it is considered equivalent to returning `false`. While it is mostly stateless (e.g. isolated from any Relay-chain state) it is provided with a `context` parameter in order to give information about a recent Relay-chain block. This allows it to be provided with a concise proof over some recent Relay-chain state.
```

Comment:
This system-wide limit was intented to ensure that its resource-requirements are modest and to imply that it would be run under metering.

Reply:
Okay, what about using a smaller proof for running authorization before sending the actual work package?

Reply:
Sure - it could be sent ahead of the WP.

Reply:
If we Merklize the WP segments and ensure that `is_authorized` only needs access to the `Authorization` segment of the WP, then sure we wouldn't need to send the whole WP before it could be authorized.

This might be a bit premature though; AFAIK we don't do anything like this currently for the collator network - indeed, we not only accept 5MB on faith, but we also allow it to execute for 2+ seconds.

Reply:
Yeah fine, I think it got probably not implemented. Tried to search for it, but couldn't find it.